### PR TITLE
typescript-msw: future proof types for generated mock functions

### PR DESCRIPTION
**Pull request by netanelgilad**
## Description

With the latest version of `msw`, the `typescript-msw` plugin generates code that doesn't compile. The generated code uses imports that are no longer available in `msw@2` (specifically `GraphQLRequest` and `GraphQLContext`).

This change makes the generated code more future proof, by typing the generated function with referencing the `typeof` for the `msw` graphql function instead of explicitly re-typing the expected `resolver`.

[See reproduction in CodeSandbox](https://codesandbox.io/p/sandbox/elastic-mclean-xn97t2?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cloo2l8h600073b6iefn4iivp%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cloo2l8h600033b6icziuqznq%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cloo2l8h600053b6iffdeg0m4%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cloo2l8h600063b6i71zurrki%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B100%252C0%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cloo2l8h600033b6icziuqznq%2522%253A%257B%2522id%2522%253A%2522cloo2l8h600033b6icziuqznq%2522%252C%2522activeTabId%2522%253A%2522cloo2on4s00v13b6iq90rzlqq%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cloo2l8h600023b6i17s2r6wm%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522id%2522%253A%2522cloo2mjpz006v3b6iga92pyg9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fcodegen.ts%2522%252C%2522id%2522%253A%2522cloo2n2xg00dx3b6illci808p%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Fcheck.tsx%2522%252C%2522id%2522%253A%2522cloo2nnrm00m03b6ifwwgobnv%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Fgql%252Fgraphql.ts%2522%252C%2522id%2522%253A%2522cloo2on4s00v13b6iq90rzlqq%2522%252C%2522mode%2522%253A%2522temporary%2522%257D%255D%257D%252C%2522cloo2l8h600063b6i71zurrki%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cloo2l8h600063b6i71zurrki%2522%257D%252C%2522cloo2l8h600053b6iffdeg0m4%2522%253A%257B%2522id%2522%253A%2522cloo2l8h600053b6iffdeg0m4%2522%252C%2522activeTabId%2522%253A%2522cloo2lbnp004c3b6il702iyfe%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cloo2l8h600043b6ifjrx6k9h%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%257D%252C%257B%2522id%2522%253A%2522cloo2lbnp004c3b6il702iyfe%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cloo2lbxj00x0efg6b6th5c6o%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Afalse%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Performed the change manually in the reproduction project.

## Checklist:

- [x] I have followed the
      [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the
      style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by dotansimha** [comment]
@netanelgilad looks great. you mentioned that the changes are future-proof, does it also mean that it's backward compatible with `msw@1`? I'm asking because if it doesn't we might want to do a major version bump to make sure it's clear that the latest generated code is generating something else (and probably also add a `peerDependency` range for v2?) 

We also need a changeset file (run `pnpm changeset` locally and it will guide you how to add it) - but let's decide if that's a breaking change that requires a major release, or just a refactor for better version support of `msw`.

**Comment by dotansimha** [comment]
@beerose any thoughts on using `typeof` in multiple places (I think it's generated per operation?)? anything we should consider regarding TS performance here?

**Comment by netanelgilad** [comment]
> does it also mean that it's backward compatible with msw@1?

yep, indeed it is (tried it out). Even ignoring the specific function/package here, since we are using `typeof` to get the type of the parameter from the actual function call, then it would always be compatible:

```typescript
const myOriginalFunction = (first: SomeType, second: AnotherType) => {}

const wrappedFunction = (second: Parameters<typeof myOriginalFunction>[1]) => { 
  return myOriginalFunction(someFirst, second);
 }
```

`myOriginalFunction` can change the type of `second` and `wrappedFunction` would always match without actually caring what `AnotherType` is.

**Comment by dotansimha** [comment]
Amazing! Can you please add a `changeset` file?

**Review by saihaj — APPROVED** [review]
Looking at the update to interface here
 - v2 https://mswjs.io/docs/api/graphql/#call-signature
 - v1 https://v1.mswjs.io/docs/api/graphql/query

seems like a safe change we are making

File: .changeset/tidy-trains-wait.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@graphql-codegen/typescript-msw': patch
```
Comment (saihaj) [inline .changeset/tidy-trains-wait.md:2]:
```suggestion
'@graphql-codegen/typescript-msw': minor
```

**Comment by saihaj** [comment]
Thank you

**Comment by saihaj** [comment]
[@graphql-codegen/typescript-msw@2.1.0](https://github.com/dotansimha/graphql-code-generator-community/releases/tag/release-1699489251768)

**Pull request merged**